clippy_lints/src/redundant_else.rs

```diff
@@ -0,0 +1,135 @@
+use crate::utils::span_lint_and_help;
+use rustc_ast::ast::{Block, Expr, ExprKind, Stmt, StmtKind};
+use rustc_ast::visit::{walk_expr, Visitor};
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for `else` blocks that can be removed without changing semantics.
+    ///
+    /// **Why is this bad?** The `else` block adds unnecessary indentation and verbosity.
+    ///
+    /// **Known problems:** Some may prefer to keep the `else` block for clarity.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// fn my_func(count: u32) {
+    ///     if count == 0 {
+    ///         print!("Nothing to do");
+    ///         return;
+    ///     } else {
+    ///         print!("Moving on...");
+    ///     }
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn my_func(count: u32) {
+    ///     if count == 0 {
+    ///         print!("Nothing to do");
+    ///         return;
+    ///     }
+    ///     print!("Moving on...");
+    /// }
+    /// ```
+    pub REDUNDANT_ELSE,
+    pedantic,
+    "`else` branch that can be removed without changing semantics"
+}
+
+declare_lint_pass!(RedundantElse => [REDUNDANT_ELSE]);
+
+impl EarlyLintPass for RedundantElse {
+    fn check_stmt(&mut self, cx: &EarlyContext<'_>, stmt: &Stmt) {
+        if in_external_macro(cx.sess, stmt.span) {
+            return;
+        }
+        // Only look at expressions that are a whole statement
+        let expr: &Expr = match &stmt.kind {
```

Comment:
The type annotations do not seem to be needed (here and below)

Reply:
The type annotation coerces `&P<Expr>` to `&Expr`. I think this is preferable to doing `&**expr` inside the match?

Reply:
Hmm, just removing the annotations compiles fine. I guess it could be related with match ergonomics being able to auto-deref if the pattern matches?

Reply:
Yes I think so.

1. type annotations - auto-deref at binding - clear and smart-pointer-agnostic
2. `&**expr` - manual deref before binding - less clear and smart-pointer-dependant
3. neither - auto-deref at each usage - simplest code but inefficient

Reply:
I think it won't make a big difference, so I think it's fine to merge it as is. Thanks!

---

To explain my reasoning:
1) I believe type annotations are generally less idiomatic if they are not needed to disambiguate (that's why I raised a concern).
2) Correct me if I'm wrong, but I would say in this case using the annotation would just change the place where the dereference happens? From the let binding to the match/function call later, but I would say the number of derefs would be the same?

Reply:
But yeah, I see your point, there would be cases where you could need more derefs than if you do it once when binding :+1: